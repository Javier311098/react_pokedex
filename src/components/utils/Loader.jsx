import React from "react";
import { CgPokemon } from "react-icons/cg";
export const Loader = () => {
  return (
    <div className="spinner-img-container">
      <CgPokemon className="img-loader" />
      <h2>Loading...</h2>
    </div>
  );
};
