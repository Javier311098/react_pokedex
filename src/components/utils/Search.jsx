import React, { useState } from "react";
import { BsSearch } from "react-icons/bs";
import { useNavigate } from "react-router-dom";
import Swal from "sweetalert2";

export const Search = () => {
  const [inputValue, setInputValue] = useState("");
  const navigate = useNavigate();

  const handleInputChange = ({ target }) => {
    if (target.value.length >= 15) {
      setInputValue("");
    } else {
      setInputValue(target.value);
    }
  };

  const handleSearch = () => {
    if (inputValue.length >= 1) {
      navigate(`/${process.env.REACT_APP_REPO_NAME}/${inputValue}`);
    } else {
      Swal.fire(
        "Error",
        "You need write a pokemon name or pokedex number",
        "error"
      );
    }
  };

  return (
    <div className="search-container">
      <div className="search-bar">
        <input
          autoComplete="false"
          type="text"
          onChange={handleInputChange}
          onKeyPress={(event) => event.key === "Enter" && handleSearch()}
          value={inputValue}
          name="inputValue"
          placeholder="Buscar pokemon..."
        />
        <button onClick={handleSearch} className="search-button">
          <BsSearch />
        </button>
      </div>
    </div>
  );
};
