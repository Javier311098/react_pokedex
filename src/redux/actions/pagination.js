import { types } from "../types/types";

export const getTotalPages = (total, limit) => {
  return (dispatch) => {
    const totalPokemons = Math.ceil(total / limit);
    dispatch(totalPages(totalPokemons));
  };
};

const totalPages = (totalPokemons) => ({
  type: types.totalPages,
  payload: totalPokemons,
});

export const currentPage = (page) => ({
  type: types.currentPage,
  payload: page,
});
