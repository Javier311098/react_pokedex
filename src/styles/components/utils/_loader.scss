.spinner-img-container {
  display: flex;
  height: 100vh;
  width: 100vw;
  align-items: center;
  justify-content: center;
  flex-direction: column;
}

.img-loader {
  width: 150px;
  height: 150px;
  animation: spin 1s ease infinite;
}

@keyframes spin {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
