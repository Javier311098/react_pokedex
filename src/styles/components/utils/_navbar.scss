.navbar-container {
  border-radius: 5px;
  border: 2px solid black;
  height: 120px;
  padding: 5px;
  margin: 5px;
  display: flex;
  align-items: center;
  justify-content: space-evenly;
  flex-direction: column;
}

.navbar-title {
  display: flex;
  align-items: center;
  img {
    width: 50px;
    height: 50px;
  }
}
