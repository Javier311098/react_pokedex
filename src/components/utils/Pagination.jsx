import React, { useEffect } from "react";
import { AiFillStepBackward, AiFillStepForward } from "react-icons/ai";
import { useDispatch } from "react-redux";
import { useSelector } from "react-redux";
import { currentPage } from "../../redux/actions/pagination";

export const Pagination = () => {
  const {
    pagination: { totalPages, actualPage },
    pokemons: { loading },
  } = useSelector((state) => state);
  const dispatch = useDispatch();

  useEffect(() => {
    const page = JSON.parse(localStorage.getItem("page") || 0);
    dispatch(currentPage(parseInt(page)));
  }, [dispatch]);

  const handleNextPage = () => {
    dispatch(currentPage(Math.min(actualPage + 1, totalPages - 1)));
    localStorage.setItem("page", Math.min(actualPage + 1, totalPages - 1));
  };

  const handleLastPage = () => {
    dispatch(currentPage(Math.max(actualPage - 1, 0)));
    localStorage.setItem("page", Math.max(actualPage - 1, 0));
  };

  return (
    <div className="pagination-container">
      <div className="pagination-container-info">
        <button
          disabled={loading ? true : false}
          className="pagination-buttons"
          onClick={handleLastPage}
        >
          <AiFillStepBackward />
        </button>
        <span>
          {actualPage + 1} of {totalPages}
        </span>
        <button
          disabled={loading ? true : false}
          className="pagination-buttons"
          onClick={handleNextPage}
        >
          <AiFillStepForward />
        </button>
      </div>
    </div>
  );
};
