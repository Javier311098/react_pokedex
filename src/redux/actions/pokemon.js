import { fetchPokemon } from "../../helpers/fetchPokemon";
import { getTotalPages } from "./pagination";
import { types } from "../types/types";
import Swal from "sweetalert2";

export const startGetPokemons = () => {
  return async (dispatch, getState) => {
    const { actualPage } = getState().pagination;
    dispatch(loadingPokemons(true));
    const limit = 40;
    const resp = await fetchPokemon(
      `pokemon?limit=${limit}&offset=${limit * actualPage}`
    );
    const body = await resp.json();
    const totalPokemons = body.count;
    const resultados = body.results;
    const pokemons = await Promise.all(
      resultados.map(async (resultado) => {
        const newResp = await fetchPokemon(`pokemon/${resultado.name}`);
        return await newResp.json();
      })
    );
    dispatch(getTotalPages(totalPokemons, limit));
    dispatch(loadingPokemons(false));
    dispatch(getPokemons(pokemons));
  };
};

const getPokemons = (pokemons) => ({
  type: types.getPokemons,
  payload: pokemons,
});

export const startGetOnePokemon = (name) => {
  return async (dispatch) => {
    try {
      dispatch(loadingPokemons(true));
      const resp = await fetchPokemon(`pokemon/${name}`);
      if (resp.status === 404) {
        Swal.fire("Error", "Could not find Pokemon", "error");
        dispatch(getError(true));
      }
      dispatch(getError(false));
      const pokemon = await resp.json();
      dispatch(pokemonSelected(pokemon));
      dispatch(loadingPokemons(false));
    } catch (error) {
      console.log(error);
    }
  };
};

export const getError = (error) => ({
  type: types.pokemonError,
  payload: error,
});

const loadingPokemons = (loading) => ({
  type: types.loadedPokemon,
  payload: loading,
});

export const resetPokemonsSelected = () => ({
  type: types.pokemonResetSelected,
});

const pokemonSelected = (pokemon) => ({
  type: types.pokemonSelected,
  payload: pokemon,
});
