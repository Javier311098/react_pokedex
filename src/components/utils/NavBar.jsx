import React from "react";
import { Pagination } from "./Pagination";
import { Search } from "./Search";

export const NavBar = () => {
  return (
    <div className="navbar-container animate__animated animate__fadeInDown">
      <div className="navbar-title">
        <h1>Pokedex</h1>
        <img
          alt="pokedex"
          src="http://i.ebayimg.com/images/g/HCgAAOxy63FSrBOH/s-l500.jpg"
        />
      </div>
      <Pagination />
      <Search />
    </div>
  );
};
