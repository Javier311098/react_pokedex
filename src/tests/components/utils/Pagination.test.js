/* eslint-disable testing-library/no-debugging-utils */

import { shallow } from "enzyme";
import { Pagination } from "../../../components/utils/Pagination";
import { currentPage } from "../../../redux/actions/pagination";

const initialState = {
  pagination: { totalPages: 28, actualPage: 2 },
  pokemons: { loading: false },
};
jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useDispatch: () => jest.fn(),
  useSelector: (selector) => selector(initialState),
}));

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useEffect: (cb) => cb(),
}));

jest.mock("../../../redux/actions/pagination", () => ({
  currentPage: jest.fn(),
}));
Storage.prototype.setItem = jest.fn();

describe("Testing Pagination ||Component", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Pagination />);
  });
  test("should render the component", () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".pagination-container").exists()).toBe(true);
  });

  test("should currentPage have called with useEffect", () => {
    expect(currentPage).toHaveBeenCalled();
  });

  test("should currentPage have called with handleNextPage save in LocalStorage", () => {
    wrapper.find("button").at(1).prop("onClick")();
    expect(localStorage.setItem).toHaveBeenCalledWith(
      "page",
      initialState.pagination.actualPage + 1
    );
    expect(currentPage).toHaveBeenCalled();
  });

  test("should currentPage have called with handleLastPage and save in LocalStorage", () => {
    wrapper.find("button").at(0).prop("onClick")();
    expect(localStorage.setItem).toHaveBeenCalledWith(
      "page",
      initialState.pagination.actualPage - 1
    );
    expect(currentPage).toHaveBeenCalled();
  });

  test("should buttons are disabled when loading its true", () => {
    jest
      .spyOn(require("react-redux"), "useSelector")
      .mockImplementation((selector) =>
        selector({
          pagination: { totalPages: 28, actualPage: 2 },
          pokemons: { loading: true },
        })
      );
    const wrapper = shallow(<Pagination />);
    expect(wrapper.find("button").at(0).prop("disabled")).toBe(true);
    expect(wrapper.find("button").at(1).prop("disabled")).toBe(true);
  });
});
