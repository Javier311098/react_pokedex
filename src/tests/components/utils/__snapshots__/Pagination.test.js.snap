// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Pagination ||Component should render the component 1`] = `
<div
  className="pagination-container"
>
  <div
    className="pagination-container-info"
  >
    <button
      className="pagination-buttons"
      disabled={false}
      onClick={[Function]}
    >
      <AiFillStepBackward />
    </button>
    <span>
      3
       of 
      28
    </span>
    <button
      className="pagination-buttons"
      disabled={false}
      onClick={[Function]}
    >
      <AiFillStepForward />
    </button>
  </div>
</div>
`;
